File: services/latr-gateway/Sources/LatrGatewayLib/Domain/RepoClient.swift
import Foundation

public struct RepoRecord<T: Codable & Sendable>: Codable, Sendable {
    public let uri: String
    public let cid: String
    public let value: T

    public init(uri: String, cid: String, value: T) {
        self.uri = uri
        self.cid = cid
        self.value = value
    }
}

public struct ListRecordsPage<T: Codable & Sendable>: Sendable {
    public let records: [RepoRecord<T>]
    public let cursor: String?

    public init(records: [RepoRecord<T>], cursor: String?) {
        self.records = records
        self.cursor = cursor
    }
}

public struct CreateRecordResult: Sendable {
    public let uri: String

    public init(uri: String) {
        self.uri = uri
    }
}

public struct PutRecordResult: Sendable {
    public let uri: String

    public init(uri: String) {
        self.uri = uri
    }
}

public protocol LatrRepoClient: Sendable {
    func listRecords<T: Codable & Sendable>(
        repo: String,
        collection: String,
        limit: Int?,
        cursor: String?
    ) async throws -> ListRecordsPage<T>

    func getRecord<T: Codable & Sendable>(
        repo: String,
        collection: String,
        rkey: String
    ) async throws -> RepoRecord<T>?

    func createRecord(
        repo: String,
        collection: String,
        rkey: String,
        record: some Encodable & Sendable
    ) async throws -> CreateRecordResult

    func putRecord(
        repo: String,
        collection: String,
        rkey: String,
        record: some Encodable & Sendable
    ) async throws -> PutRecordResult

    func deleteRecord(
        repo: String,
        collection: String,
        rkey: String
    ) async throws
}

File: packages/latr-kit/src/repoClient.ts
/** ATProto repo record returned from list/get XRPC. */
export interface RepoRecord<T> {
  uri: string;
  cid: string;
  value: T;
}

export interface CreateRecordResult {
  uri: string;
}

export interface PutRecordResult {
  uri: string;
}

export interface ListRecordsPage<T> {
  records: RepoRecord<T>[];
  cursor?: string;
}

/**
 * Protocol-neutral repo client for L@tr write workflows.
 * Apps adapt OAuth-backed `@atproto/api` Agent implementations to this interface.
 */
export interface LatrRepoClient {
  listRecords<T>(params: {
    repo: string;
    collection: string;
    limit?: number;
    cursor?: string;
  }): Promise<ListRecordsPage<T>>;

  getRecord<T>(params: {
    repo: string;
    collection: string;
    rkey: string;
  }): Promise<RepoRecord<T> | null>;

  createRecord(params: {
    repo: string;
    collection: string;
    rkey: string;
    record: Record<string, unknown>;
  }): Promise<CreateRecordResult>;

  putRecord(params: {
    repo: string;
    collection: string;
    rkey: string;
    record: Record<string, unknown>;
  }): Promise<PutRecordResult>;

  deleteRecord(params: {
    repo: string;
    collection: string;
    rkey: string;
  }): Promise<void>;
}
